crates/proto/src/dnssec/dnssec_dns_handle/mod.rs

```diff
@@ -775,6 +767,13 @@ fn verify_dnskey(
             continue;
         }
 
+        key_authentication_attempts += 1;
+
+        if key_authentication_attempts > MAX_KEY_TAG_COLLISIONS {
+            warn!("too many DS records ({key_authentication_attempts}) with key tag {key_tag}; skipping");
```

Comment:
Take the opportunity to use structured logging for this?

I think you should add a comment making it clear that the ordering of (1) comparing algorithms, (2) rate limiting is load-bearing.